apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ElasticApmTracer.java

```diff
@@ -193,6 +201,18 @@ public void onChange(ConfigurationOption<?> configurationOption, Double oldValue
         return startTransaction(childContextCreator, parent, sampler, -1, initiatingClassLoader);
     }
 
+    public void avoidWrappingOnThread() {
```

Comment:
I think we should make it so that it also works when things are wrapped multiple times. Avoiding the wrapping as you did is ok but should "just" be an optimization. It should not be required to call these methods.

Reply:
I agree, which is why I added another guard where if wrapped twice, same span won't be activated and deactivated twice using this flag: https://github.com/elastic/apm-agent-java/pull/605/files#diff-d1e413fd7a24f92f13f6094cbf2dd787R37

In addition, I am relying on your elimination of https://github.com/elastic/apm-agent-java/blob/master/apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ElasticApmTracer.java#L478, which, if moved, things would not be optimal but work properly. This is the only thing that currently makes multiple activation/deactivation of a span not fully safe.

Reply:
Could you remove that in your pr as well? Reason is that we probably want to release this PR with a bug fix release soon whereas the other PR will be released later in a feature release.

Reply:
Sure, good point.